modules/product/src/main/java/com/opengamma/strata/product/cms/Cms.java

```diff
@@ -0,0 +1,427 @@
+/**
+ * Copyright (C) 2015 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.product.cms;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.Set;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.ImmutableValidator;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.product.swap.SwapIndex;
+import com.opengamma.strata.product.swap.SwapLeg;
+
+/**
+ * The class defines a constant maturity swap (CMS) or CMS cap/floor. 
+ * <p>
+ * The CMS product consists of two legs: CMS leg and pay leg. 
+ * The CMS leg of CMS periodically pays coupons based on swap rate, the observed value of {@linkplain SwapIndex swap index},  
+ * CMS cap/floor is a set of call/put options on successive swap rates, i.e., CMS caplets/floorlets. 
+ * The other leg is typically the same as a swap leg of the standard interest rate swap. See {@link SwapLeg}.
+ * <p>
+ * However, the pay leg is absent for certain CMS products. Instead the premium is paid upfront. See {@link CmsTrade}.
+ */
+@BeanDefinition
+public final class Cms
+    implements CmsProduct, ImmutableBean, Serializable {
+
+  /**
+   * The CMS leg of the product.
+   * <p>
+   * This is associated with periodic payments based on swap rate. 
+   * The payments are CMS coupons, CMS caplets or CMS floors. 
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final CmsLeg cmsLeg;
+  /**
+   * The pay leg of the product. 
+   * <p>
+   * Typically this is associated with periodic fixed or Ibor rate payments without compounding or notioanl exchange. 
+   * <p>
+   * These periodic payments are not made over the lifetime of the product for certain CMS products. Instead the 
+   * premium is paid upfront. 
+   */
+  @PropertyDefinition(get = "optional")
+  private final SwapLeg payLeg;
+
+  //-------------------------------------------------------------------------
```

Comment:
As there are only 2 properties, it is probably worth adding an `of(CmsLeg, SwapLeg)` static method.